build/lib/src/state/input_tracker.dart

```diff
@@ -6,7 +6,30 @@ import 'dart:collection';
 
 import '../asset/id.dart';
 
-/// Records inputs for one generator run.
+/// Records inputs for each generator.
 class InputTracker {
-  final assetsRead = HashSet<AssetId>();
+  final _inputs = HashMap<AssetId, HashSet<AssetId>>();
```

Comment:
I don't believe this is safe to do this way, if I understand it correctly. The keys in this map would need to be the output ids, instead of the primary input id.

The reason being that an asset can be a primary input to any number of different actions.

I would try adding a test where there are two builders with the same input extension, but which read different additional files during the build, and make sure they don't pick up each other's dependencies.

Reply:
Another consideration here is this will retain these sets much longer in memory (for an entire build), instead of releasing them as soon as a build completes.

Reply:
I think there is a much simpler answer here, if the state is per `BuildStep`, it should be stored per `BuildStep` :) ... I think I can get that working.